### ISSUE TITLE
mvnw build in Ubuntu with container-image-s2i successful but did not generate OpenShift manifests

### ISSUE THREAD

**Issue by jackhu1** [quarkusio/quarkus#11730]
**Describe the bug**
(Describe the problem clearly and concisely.)
Following on from resolution of issue #11539, I wanted to use s2i to containerise and deploy my quarkus-greeting-app to OpenShift.  Ref https://quarkus.io/guides/container-image#s2i.

I added s2i container-image extension to my project and did a clean build:

`jhu120@DESKTOP-C7JTV74:~/quarkus_java/quarkus-greeting-app$ ./mvnw quarkus:add-extension -Dextensions="container-image-s2i"`

`jhu120@DESKTOP-C7JTV74:~/quarkus_java/quarkus-greeting-app$  ./mvnw clean package  -Dquarkus.container-image.build=true -Dquarkus.container-image.builder=s2i`

Build completed successfully but it did not generate an OpenShift manifest.  I also tried adding "-Dquarkus.kubernetes.deployment-target=openshift " to the build command without luck:

`jhu120@DESKTOP-C7JTV74:~/quarkus_java/quarkus-greeting-app$ ./mvnw clean package  -Dquarkus.container-image.build=true -Dquarkus.container-image.builder=s2i -Dquarkus.kubernetes.deployment-target=openshift`

Ref: https://quarkus.io/guides/deploying-to-kubernetes

Environment:
Ubuntu WSL 16.04 for Windows 10
Oracle JDK 11.0.8 
Maven 3.6.3 (using Oracle Java 10.0.8 /usr/lib/jvm/java-11-oracle)

**Console output**
```
jhu120@DESKTOP-C7JTV74:~/quarkus_java/quarkus-greeting-app$ ./mvnw clean package  -Dquarkus.container-image.build=true -Dquarkus.container-image.builder=s2i
[INFO] Scanning for projects...
[INFO]
[INFO] ------------< io.openshifteverywhere:quarkus-greeting-app >-------------
[INFO] Building quarkus-greeting-app 1.0-SNAPSHOT
[INFO] --------------------------------[ jar ]---------------------------------
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-clean-plugin/2.5/maven-clean-plugin-2.5.pom
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-clean-plugin/2.5/maven-clean-plugin-2.5.pom (3.9 kB at 4.8 kB/s)
Downloading from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-clean-plugin/2.5/maven-clean-plugin-2.5.jar
Downloaded from central: https://repo.maven.apache.org/maven2/org/apache/maven/plugins/maven-clean-plugin/2.5/maven-clean-plugin-2.5.jar (25 kB at 57 kB/s)
[INFO]
[INFO] --- maven-clean-plugin:2.5:clean (default-clean) @ quarkus-greeting-app ---
[INFO] Deleting /home/jhu120/quarkus_java/quarkus-greeting-app/target
[INFO]
[INFO] --- quarkus-maven-plugin:1.7.0.Final:prepare (default) @ quarkus-greeting-app ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:resources (default-resources) @ quarkus-greeting-app ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] Copying 2 resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:compile (default-compile) @ quarkus-greeting-app ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 3 source files to /home/jhu120/quarkus_java/quarkus-greeting-app/target/classes
[INFO]
[INFO] --- quarkus-maven-plugin:1.7.0.Final:prepare-tests (default) @ quarkus-greeting-app ---
[INFO]
[INFO] --- maven-resources-plugin:2.6:testResources (default-testResources) @ quarkus-greeting-app ---
[INFO] Using 'UTF-8' encoding to copy filtered resources.
[INFO] skip non existing resourceDirectory /home/jhu120/quarkus_java/quarkus-greeting-app/src/test/resources
[INFO]
[INFO] --- maven-compiler-plugin:3.8.1:testCompile (default-testCompile) @ quarkus-greeting-app ---
[INFO] Changes detected - recompiling the module!
[INFO] Compiling 2 source files to /home/jhu120/quarkus_java/quarkus-greeting-app/target/test-classes
[INFO]
[INFO] --- maven-surefire-plugin:3.0.0-M5:test (default-test) @ quarkus-greeting-app ---
[INFO]
[INFO] -------------------------------------------------------
[INFO]  T E S T S
[INFO] -------------------------------------------------------
[INFO] Running io.openshifteverywhere.greeting.app.GreetingResourceTest
2020-08-30 00:11:29,798 INFO  [io.quarkus] (main) Quarkus 1.7.0.Final on JVM started in 3.590s. Listening on: http://0.0.0.0:8081
2020-08-30 00:11:29,801 INFO  [io.quarkus] (main) Profile test activated.
2020-08-30 00:11:29,801 INFO  [io.quarkus] (main) Installed features: [cdi, resteasy]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0, Time elapsed: 11.006 s - in io.openshifteverywhere.greeting.app.GreetingResourceTest
2020-08-30 00:11:33,984 INFO  [io.quarkus] (main) Quarkus stopped in 0.232s
[INFO]
[INFO] Results:
[INFO]
[INFO] Tests run: 2, Failures: 0, Errors: 0, Skipped: 0
[INFO]
[INFO]
[INFO] --- maven-jar-plugin:2.4:jar (default-jar) @ quarkus-greeting-app ---
[INFO] Building jar: /home/jhu120/quarkus_java/quarkus-greeting-app/target/quarkus-greeting-app-1.0-SNAPSHOT.jar
[INFO]
[INFO] --- quarkus-maven-plugin:1.7.0.Final:build (default) @ quarkus-greeting-app ---
[INFO] [org.jboss.threads] JBoss Threads version 3.1.1.Final
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building thin jar: /home/jhu120/quarkus_java/quarkus-greeting-app/target/quarkus-greeting-app-1.0-SNAPSHOT-runner.jar
[WARNING] [io.quarkus.container.image.s2i.deployment.S2iProcessor] No Openshift manifests were generated (most likely due to the fact that the service is not an HTTP service) so no s2i process will be taking place
[INFO] [io.quarkus.deployment.QuarkusAugmentor] Quarkus augmentation completed in 3464ms
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  51.138 s
[INFO] Finished at: 2020-08-30T00:11:38+01:00
[
```

**Expected behavior**
(Describe the expected behavior clearly and concisely.)
The build should have created two manifests: 
openshift.json and openshift.yml in the target/kubernetes/ but it did not.

**Actual behavior**
(Describe the actual behavior clearly and concisely.)
 Folder ./target/kubernetes was not created.

**To Reproduce**
1. see the two commands in the bug description.
2. 
3. 

**Configuration**
```properties
# Add your application.properties here, if applicable.
N/A. I passed properties on the command line.
```

**Screenshots**
(If applicable, add screenshots to help explain your problem.)

**Environment (please complete the following information):**
 - Output of `uname -a` or `ver`: 
 - Output of `java -version`: 
 - GraalVM version (if different from Java): 
 - Quarkus version or git rev: 
 - Build tool (ie. output of `mvnw --version` or `gradlew --version`): 

**Additional context**
(Add any other context about the problem here.)

**Comment by geoand** [comment]
I just tried this, and adding `-Dquarkus.kubernetes.deployment-target=openshift` works as expected.

Do you have a reproducer project that exhibits the behavior you mention?

**Comment by jackhu1** [comment]
Hello @geoand , thanks for your quick reply!  I managed to create the OpenShift manifest files after modifying my procedures.

Here are details.
I first tried the tutorial  https://quarkus.io/guides/deploying-to-kubernetes#openshift from scratch, i.e. run the bootstrapping command:

```
mvn io.quarkus:quarkus-maven-plugin:1.7.1.Final:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=kubernetes-quickstart \
    -DclassName="org.acme.rest.GreetingResource" \
    -Dpath="/greeting" \
    -Dextensions="kubernetes, jib"
```
and then build with 

`./mvnw clean package  -Dquarkus.container-image.builder=s2i -Dquarkus.kubernetes.deployment-target=openshift `

and it worked, successfully created openshift.json ad openshift.yml files in target/kubernetes/.

I then compared this with my own application. My own application is a customised version of https://quarkus.io/guides/getting-started.  I can see that my own application does not have the extension "-Dextensions="kubernetes, jib".  

After some trial and error, I managed to create the OpenShift manifests with the following commands against my own application:

`~/quarkus_java/quarkus-greeting-app$ ./mvnw quarkus:add-extension -Dextensions="openshift"`  I added "openshift" extension instead of "kubernetes, jib" extensions to my project pom.xml via this command. 

`~/quarkus_java/quarkus-greeting-app$ ./mvnw clean package  -Dquarkus.container-image.builder=s2i -Dquarkus.kubernetes.deployment-target=openshift `  Note that I removed "-Dquarkus.container-image.build=true" from the arguments as no image is required to be built locally when using OpenShift s2i.

These modifications worked and created OpenShift manifest files openshift.yml and openshift.json  in target/kubernetes/.

**Comment by jackhu1** [comment]
Then I encountered another issue trying to deploy the app into OpenShift - just like peeling an onion :-( :
`oc apply -f ./target/kubernetes/openshift_8080.yml` created resources but did not try to build an application image.
Tried manual bc build which completed but deployment failed with error "Error: Unable to access jarfile /deployments/quarkus-greeting-1.0-SNAPSHOT-runner.jar"

Details of the issue:

```
~/quarkus_java/quarkus-greeting-app$ oc apply -f ./target/kubernetes/openshift_8088.yml
serviceaccount/quarkus-greeting-app unchanged
service/quarkus-greeting-app created
imagestream.image.openshift.io/openjdk-11 unchanged
imagestream.image.openshift.io/quarkus-greeting-app unchanged
buildconfig.build.openshift.io/quarkus-greeting-app created
deploymentconfig.apps.openshift.io/quarkus-greeting-app created
```
Note: changed port to 8088 from 8080 to avoid conflict with an existing quarkus running pod on port 8080, which was deployed from an imaged built with Docker build.

Status showed bc never built (repeatedly run the below query):
```
~/quarkus_java/quarkus-greeting-app$ oc status
...
svc/quarkus-greeting-app - 172.21.131.75:8088
  dc/quarkus-greeting-app deploys istag/quarkus-greeting-app:1.0-SNAPSHOT <-
    bc/quarkus-greeting-app source builds uploaded code on istag/openjdk-11:latest
      not built yet
    deployment #1 waiting on image
```
Query pods showed no build activities for quarkus-greeting-app :
```
~/quarkus_java/quarkus-greeting-app$ oc get pods -w
NAME                           READY     STATUS      RESTARTS   AGE
...
<no quarkus-greeting-app-1-deploy listed>
```
Query image stream showed image stream quarkus-greeting-app was present but missing a tag. 
:

```
~/quarkus_java/quarkus-greeting-app$ oc get is
NAME                   IMAGE REPOSITORY                                                                   TAGS                                     UPDATED
openjdk-11             registry.access.redhat.com/ubi8/openjdk-11                                         1.3,1.3-3,1.3-3.1591609340 + 6 more...   3 hours ago
quarkus-greeting       image-registry.openshift-image-registry.svc:5000/jackchu-uk/quarkus-greeting       1.0-SNAPSHOT                             20 minutes ago
quarkus-greeting-app   image-registry.openshift-image-registry.svc:5000/jackchu-uk/quarkus-greeting-app
j
```

So I tried a manual build: 

`oc start-build bc/quarkus-greeting-app`

The manual build completed but deployment failed:

```
~/quarkus_java/quarkus-greeting-app$ oc get pods
NAME                            READY     STATUS             RESTARTS   AGE
...
quarkus-greeting-app-1-build    0/1       Completed          0          2m7s
quarkus-greeting-app-1-deploy   1/1       Running            0          98s
quarkus-greeting-app-1-qflfs    0/1       CrashLoopBackOff   3          97s

```
The failed pod shows the following error:  
```
~/quarkus_java/quarkus-greeting$ oc logs quarkus-greeting-1-qflfs    
Error: Unable to access jarfile /deployments/quarkus-greeting-1.0-SNAPSHOT-runner.jar
```

**Comment by geoand** [comment]
The generated Openshift manifests assume that a binary s2i build is going to take place. That is why letting Quarkus orchestrate the process is the best way to go.

So you generally use `mvn package -Dquarkus.kubernetes.deploy=true` and Quarkus will apply the manifests and then stgart the proper s2i binary build.

I am going to close this, as I don't see any Quarkus issue here - but feel free to reopen and / or comment further :)

**Comment by jackhu1** [comment]
Hello @geoand,  well, my experience so far suggests that there is a documentation issue.  It has been challenging to follow through the documentation to use s2i extension to deploy the sample application to OpenShift.   Following instructions in https://quarkus.io/guides/deploying-to-kubernetes#openshift and combining it with instructions in https://quarkus.io/guides/container-image#s2i  has not lead to a successful deployment to OpenShift using s2i.

**Comment by geoand** [comment]
Looking at https://quarkus.io/guides/deploying-to-openshift, it seems to be fine, but we always welcome doc updates :)

**Comment by jackhu1** [comment]
Hello @geoand, thank you so much for your last comment!  Awesome. The instructions there worked and I managed to deploy the application to OpenShift using the implied s2i binary build.

Two feedback comments for improvement of the documentation:

1. After the build (using s2i to create image stream) only command `./mvnw clean package -Dquarkus.container-image.build=true`, it would be helpful to show instructions to deploy the created image stream to OpenShift.  
I used the basic command `oc new-app --name=greeting jackchu-quarkus/openshift-quickstart:1.0-SNAPSHOT` to deploy.  The complete set of commands that I used were:
```
oc get is
oc new-app --name=greeting jackchu-quarkus/openshift-quickstart:1.0-SNAPSHOT
oc get svc
oc expose svc/greeting
oc get route
curl http://<route>/greeting
```
2. Revise the instructions relating to building for and deploying to OpenShift in https://quarkus.io/guides/deploying-to-kubernetes
That article describes commands for deploying to generate OpenShift resources and to deploy to OpenShift. Those did not work for me and they are different from those in https://quarkus.io/guides/deploying-to-openshift.
Maybe someone could go through those commands to see what changes are needed to make them work.

Many thanks for patiently staying with me and answering my questions.  You have been very helpful! Jack

**Comment by jackhu1** [comment]
I will go back to my original application quarkus-greeting-app and try the new command to make it work.  I will post an update here for information.

**Comment by geoand** [comment]
@jackhu1 glad it worked!

Do you fancy sending a PR for the doc updates?
It's often the case that a new perspective brings much needed doc improvements

**Comment by jackhu1** [comment]
Yes, happy to do.  Need a bit guidance about the procedures though.  Would you mind pointing me in the right direction? Perhaps starting with where to find the document source to edit.
  
I understand the concepts of code branching, code pull and push and raising PR but need to know the whereabout of  Quarkus documents and the specific procedures to make changes.

**Comment by geoand** [comment]
Sure thing.

All our docs are here: https://github.com/quarkusio/quarkus/tree/master/docs/src/main/asciidoc

So basically all you need to do is fork the project, edit the files, push the changes and then open a GitHub Pull Request against the Quarkus repo. 
You can also check out https://github.com/quarkusio/quarkus/blob/master/CONTRIBUTING.md for more information

**Comment by jackhu1** [comment]
Hello @geoand great.  Will do (an evening job probably).

**Comment by jackhu1** [comment]
> I will go back to my original application quarkus-greeting-app and try the new command to make it work. I will post an update here for information.

Updated:  tried the two stepped method 1) on my original app quarkus-greeting-app and it worked! i.e.
```
./mvnw clean package -Dquarkus.container-image.build=true
oc new-app --name=greeting-s2i jackchu-uk/quarkus-greeting-app:1.0-SNAPSHOT 
```
Not tried method 2) yet but pretty sure it would work.

**Comment by jackhu1** [comment]
Hi @geoand,  sorry I have been non-absent for a while. I am checking with my company to see what clearance process that I need to go through before I can contribute to the documentation.

**Comment by geoand** [comment]
Cool, thanks for the update

**Comment by jackhu008** [comment]
Hi @geoand, I have completed my company's process for contributing to this open source project. 
I have folked the repository, modified the document "Deploy to OpenShift" and raised a PR.
I cannot see an option for adding a reviewer. So I added your name in the comment.

**Comment by jackhu008** [comment]
Will modify https://quarkus.io/guides/deploying-to-kubernetes regarding deploying to OpenShift part another time and raise a separate PR.
